lib/session.js

```diff
@@ -0,0 +1,102 @@
+'use strict';
+
+const uuid4 = require('uuid/v4');
+
+// JSTP Session class used to buffer and resend the messages on unexpected
+// connection closes.
+// Extends Map class and thus can be used to store the current session state
+// independently of connection.
+//
+class Session extends Map {
+  //   connection - Connection class object
+  //   sessionId - used on client to represent session with corresponding id
+  //               which exists on the server
+  //   state - Map object that can be provided to initialize session state
+  //
+  constructor(connection, sessionId, state) {
+    super(state);
+
+    this.id = sessionId || uuid4();
+    this.connection = connection;
+    this.username = null;
+
+    this.guaranteedDeliveredCount = 0;
+    this.buffer = new Map();
+
+    this.recievedCount = 0;
+    this.latestBufferedMessageId = 0;
+
+    Object.preventExtensions(this);
```

Comment:
To prevent users from creating new fields in session objects and thus enforce users to use session object's storage only by using methods inherited from Map.